### PR TITLE
Adding HZZ electron MVA ID

### PR DESCRIPTION
**Pull request by asculac**
#### PR description:
- adding HZZ training for electron ID, which is an EGgamma-approved ID (latest talk at egamma POG [here](https://indico.cern.ch/event/895972/contributions/4297090/attachments/2219102/3757479/ElectronID_Egamma31032021.pdf) ) used for H->4l, which cannot otherwise be recomputed on top of nanoAODs. 
- It is based on the same electronMVAValueMapProducer used for existing mvaIds, but with an updated and separate training for 2016UL, 2017UL and 2018UL.
- A single variable mvaHZZIso is added to electrons. The appropriate training to be used is selected automatically. The idea is that the variable name will stay the same in the future, selecting a new training when appropriate.
- No need to store working points as bools since these can be easily derived from the MVA value.

- In addition to this PR: renaming mva Egamma IDs to be more generic 

#### PR validation:

- Tested in CMSSW_12_4_0_pre1; verified that the new variable is added correctly.
- Adds 1 float per electron.

@namapane @swagata87  pls follow

*this PR is a repeated clean version of previous PR (https://github.com/cms-sw/cmssw/pull/37429)

### CODE REVIEW

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
@asculac
can you prepare in the mean time the backport to the 10_4_X ?

File: PhysicsTools/NanoAOD/python/electrons_cff.py [hunk 1]
```diff
@@ -305,6 +309,15 @@ def _get_bitmapVIDForEle_docstring(modules,WorkingPoints):
                       VIDNestedWPBitmapSpring15 = cms.InputTag("bitmapVIDForEleSpring15"),
                       VIDNestedWPBitmapSum16 = cms.InputTag("bitmapVIDForEleSum16"),
                       )
+(~run2_nanoAOD_preUL & run2_egamma_2016).toModify(slimmedElectronsWithUserData.userFloats,
+    mvaHZZIso = cms.InputTag("electronMVAValueMapProducer:ElectronMVAEstimatorRun2Summer16ULIdIsoValues"),
```
Comment (perrotta) [inline PhysicsTools/NanoAOD/python/electrons_cff.py:313]:
```suggestion
    mvaHZZIso = "electronMVAValueMapProducer:ElectronMVAEstimatorRun2Summer16ULIdIsoValues", 
```

File: PhysicsTools/NanoAOD/python/electrons_cff.py [hunk 2]
```diff
@@ -305,6 +309,15 @@ def _get_bitmapVIDForEle_docstring(modules,WorkingPoints):
                       VIDNestedWPBitmapSpring15 = cms.InputTag("bitmapVIDForEleSpring15"),
                       VIDNestedWPBitmapSum16 = cms.InputTag("bitmapVIDForEleSum16"),
                       )
+(~run2_nanoAOD_preUL & run2_egamma_2016).toModify(slimmedElectronsWithUserData.userFloats,
+    mvaHZZIso = cms.InputTag("electronMVAValueMapProducer:ElectronMVAEstimatorRun2Summer16ULIdIsoValues"), 
+)
+(~run2_nanoAOD_preUL & run2_egamma_2017).toModify(slimmedElectronsWithUserData.userFloats,
+    mvaHZZIso = cms.InputTag("electronMVAValueMapProducer:ElectronMVAEstimatorRun2Summer17ULIdIsoValues") ,
```
Comment (perrotta) [inline PhysicsTools/NanoAOD/python/electrons_cff.py:316]:
```suggestion
    mvaHZZIso = "electronMVAValueMapProducer:ElectronMVAEstimatorRun2Summer17ULIdIsoValues" ,
```

File: PhysicsTools/NanoAOD/python/electrons_cff.py [hunk 3]
```diff
@@ -305,6 +309,15 @@ def _get_bitmapVIDForEle_docstring(modules,WorkingPoints):
...
+)
+(~run2_nanoAOD_preUL & run2_egamma_2018).toModify(slimmedElectronsWithUserData.userFloats,
+    mvaHZZIso = cms.InputTag("electronMVAValueMapProducer:ElectronMVAEstimatorRun2Summer18ULIdIsoValues"),
```
Comment (perrotta) [inline PhysicsTools/NanoAOD/python/electrons_cff.py:319]:
```suggestion
    mvaHZZIso = "electronMVAValueMapProducer:ElectronMVAEstimatorRun2Summer18ULIdIsoValues", 
```

**Comment by mariadalfonso** [comment]
Similarly to  what @perrotta pointed out, there are more places where the same cleanup is useful
https://github.com/cms-sw/cmssw/blob/e0f0c1c30d64cc64c3bc0b42c2bcfe6bb56e4f5d/PhysicsTools/NanoAOD/python/electrons_cff.py#L38
https://github.com/cms-sw/cmssw/blob/e0f0c1c30d64cc64c3bc0b42c2bcfe6bb56e4f5d/PhysicsTools/NanoAOD/python/electrons_cff.py#L164-L191
https://github.com/cms-sw/cmssw/blob/e0f0c1c30d64cc64c3bc0b42c2bcfe6bb56e4f5d/PhysicsTools/NanoAOD/python/electrons_cff.py#L269-L310

**Comment by mariadalfonso** [comment]
@asculac
one more requests: the renaming of the FallV2 ID should be also done in the DQM
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/python/nanoDQM_cfi.py#L86-L93

**Comment by clacaputo** [comment]
@cmsbuild please test

Just refreshing the test results

**Comment by mariadalfonso** [comment]
@asculac
can you tell when when we finalise the remaining items from the review ?
i.e. DQM + various removal of declaration

**Comment by asculac** [comment]
> @asculac can you prepare in the mean time the backport to the 10_4_X ?

Is it okay for backport to be without cleanup (declaration removal and DQM) or should I also include it?

**Comment by clacaputo** [comment]
@cmsbuild please test

File: PhysicsTools/NanoAOD/python/electrons_cff.py [hunk 4]
```diff
@@ -264,47 +268,56 @@ def _get_bitmapVIDForEle_docstring(modules,WorkingPoints):
...
-    cutbasedID_Spring15_veto = cms.InputTag("egmGsfElectronIDs:cutBasedElectronID-Spring15-25ns-V1-standalone-veto"),
-    cutbasedID_Spring15_loose = cms.InputTag("egmGsfElectronIDs:cutBasedElectronID-Spring15-25ns-V1-standalone-loose"),
-    cutbasedID_Spring15_medium = cms.InputTag("egmGsfElectronIDs:cutBasedElectronID-Spring15-25ns-V1-standalone-medium"),
-    cutbasedID_Spring15_tight = cms.InputTag("egmGsfElectronIDs:cutBasedElectronID-Spring15-25ns-V1-standalone-tight"),
+    mvaSpring16GP_WP90 = "egmGsfElectronIDs:mvaEleID-Spring16-GeneralPurpose-V1-wp90",
+    mvaSpring16GP_WP80 = "egmGsfElectronIDs:mvaEleID-Spring16-GeneralPurpose-V1-wp80",
+    mvaSpring16HZZ_WPL = "egmGsfElectronIDs:mvaEleID-Spring16-HZZ-V1-wpLoose",
+    cutbasedID_Sum16_veto = "egmGsfElectronIDs:cutBasedElectronID-Summer16-80X-V1-veto",
+    cutbasedID_Sum16_loose = "egmGsfElectronIDs:cutBasedElectronID-Summer16-80X-V1-loose",
+    cutbasedID_Sum16_medium = ("egmGsfElectronIDs:cutBasedElectronID-Summer16-80X-V1-medium",
```
Comment (clacaputo) [inline PhysicsTools/NanoAOD/python/electrons_cff.py:298]:
```suggestion
    cutbasedID_Sum16_medium = "egmGsfElectronIDs:cutBasedElectronID-Summer16-80X-V1-medium",
```
@asculac this should fix the compilation error, considering that the log states

```
*** Error compiling 'src/PhysicsTools/NanoAOD/python/electrons_cff.py'...
  File "src/PhysicsTools/NanoAOD/python/electrons_cff.py", line 299
    cutbasedID_Sum16_tight = "egmGsfElectronIDs:cutBasedElectronID-Summer16-80X-V1-tight",
                           ^
SyntaxError: invalid syntax
```

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
Error is as follow
```
  File "/cvmfs/cms-ib.cern.ch/nweek-02739/el8_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-06-29-2300/src/FWCore/ParameterSet/python/Config.py", line 1725, in _raiseUnknownKey
    raise KeyError("Unknown parameter name "+key+" specified while calling Modifier")
KeyError: 'Unknown parameter name mvaSpring16GP specified while calling Modifier'
```

**Comment by asculac** [comment]
> Error is as follow
> 
> ```
>   File "/cvmfs/cms-ib.cern.ch/nweek-02739/el8_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-06-29-2300/src/FWCore/ParameterSet/python/Config.py", line 1725, in _raiseUnknownKey
>     raise KeyError("Unknown parameter name "+key+" specified while calling Modifier")
> KeyError: 'Unknown parameter name mvaSpring16GP specified while calling Modifier'
> ```

@mariadalfonso I am not sure why this error is raised since this is the cleanup related change, I could put back the cms.InputTag but I don't see why this happens for mvaSpring16GP case

**Comment by mariadalfonso** [comment]
> > Error is as follow
> > ```
> >   File "/cvmfs/cms-ib.cern.ch/nweek-02739/el8_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-06-29-2300/src/FWCore/ParameterSet/python/Config.py", line 1725, in _raiseUnknownKey
> >     raise KeyError("Unknown parameter name "+key+" specified while calling Modifier")
> > KeyError: 'Unknown parameter name mvaSpring16GP specified while calling Modifier'
> > ```
> 
> @mariadalfonso I am not sure why this error is raised since this is the cleanup related change, I could put back the cms.InputTag but I don't see why this happens for mvaSpring16GP case

@asculac 

The inputTag need to defined at least once and then for each replacement we can avoid to redeclare.
I think this mvaSpring16GP is used here for the first time and doesn't appear in the main declaration.
https://github.com/cms-sw/cmssw/blob/127e286cdd806c161fde8f919ca86f9cbbfeca0b/PhysicsTools/NanoAOD/python/electrons_cff.py#L195

I suggest to add back, for this specific case.
runTheMatrix -l 1329.1
runTheMatrix -l 136.7722
before push

**Comment by mariadalfonso** [comment]
> Regarding the backport, should I prepare it with all of the changes (adding also cleanup) or only changes related to adding HZZ electron mva ID?

easier to do 1:1 backport

**Comment by asculac** [comment]
@mariadalfonso 

I am confused as I don't see a specific case for which the code is breaking now and it has passed runTheMatrix locally for me) with two problematic cases from before. I see only that it has timed out after 4 hours with no specific error message, am I missing something?

**Comment by mariadalfonso** [comment]
Still some error when we run w/ modifier 94X2016
KeyError: 'Unknown parameter name cutbasedID_Sum16_veto specified while calling Modifier'

**Comment by mariadalfonso** [comment]
@asculac
can you take care of this asap ? it's becoming urgent as we need to start the  run3 nano asap.

**Comment by asculac** [comment]
@mariadalfonso 

> Still some error when we run w/ modifier 94X2016 KeyError: 'Unknown parameter name cutbasedID_Sum16_veto specified while calling Modifier'

I don't see this error, I see that all checks have passed? 
I apologise for the delay with this PR, I will wrap it up today.
If everything is good here I have a question about the backport. Is it okay to do it on 10_6_X since there I have the backport of adding HZZ UL ID https://github.com/cms-sw/cmssw/pull/33620 ? So that I don't have to re-add it in 10_4_X

**Comment by mariadalfonso** [comment]
94X2016 is not running in the runTheMatrix, this is why you do not see here 
but here https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/issues/153

Anyhow you need to roll back the changes of the InputTag for the 94X2016 modifier as done previously for the 8X.

for the backport we need in 12_4[
sorry for the typo creating confusion here
https://github.com/cms-sw/cmssw/pull/38355#issuecomment-1158576889 ]

**Comment by asculac** [comment]
> 94X2016 is not running in the runTheMatrix, this is why you do not see here but here https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/issues/153
> 
> Anyhow you need to roll back the changes of the InputTag for the 94X2016 modifier as done previously for the 8X.
> 
> for the backport we need in 12_4[ sorry for the typo creating confusion here https://github.com/[/pull/38355](https://github.com/cms-sw/cmssw/pull/38355)#issuecomment-1158576889 ]

Thank you for clearing it up. If this passes tests I'll immediately do the backport for 12_4_X

**Comment by clacaputo** [comment]
@cmsbuild please test

**Comment by gouskos** [comment]
+xpog
Modifications in line with PR description

**Comment by clacaputo** [comment]
+reconstruction

- RECO differences are related to the new variable "mvaHZZIso", not present in the reference release. This new var shifts some electronTable columns, introducing some artefact differences. In the table below you can see that the values of the already present variables (although the change in name) are the same

| This PR | Base ref|
|---|---|
|miniPFRelIso_chg     | miniPFRelIso_chg |
|1.8108803033828735   | 1.8108803033828735 |
|mvaHZZIso            | **mvaFall17V2Iso** |
|-1.0                 | **-0.999984622001648** |
|**mvaIso**               | mvaFall17V2Iso_WP80 |
|**-0.999984622001648**   | 0.0 |
|mvaIso_WP80          | mvaFall17V2Iso_WP90 |
|0.0                  | 0.0 |
|mvaIso_WP90          | mvaFall17V2Iso_WPL |
|0.0                  | 0.0 |
|mvaIso_WPL           | mvaFall17V2noIso |
|0.0                  | -0.9999901056289673 |
|mvaNoIso             | mvaFall17V2noIso_WP80 |
|-0.9999901056289673  | 0.0 |
|mvaNoIso_WP80        | mvaFall17V2noIso_WP90 |
|0.0                  | 0.0 |
|mvaNoIso_WP90        | mvaFall17V2noIso_WPL |
|0.0                  | 0.0 |

**Comment by perrotta** [comment]
+1

**Pull request merged**